ARROW-9304: [C++] Add "AppendEmpty" builder APIs for use inside StructBuilder::AppendNull

tianchen92 opened the pull request.

pitrou commented on cpp/src/arrow/buffer_builder.h:
```diff
@@ -292,6 +292,11 @@ class TypedBufferBuilder<bool> {
     return Status::OK();
   }
 
+  void AppendEmpty(int64_t num_elements) {
```
We must not leave elements undefined in an array, because it can reveal previous data. "Empty" values here should still be non-null values of arbitrary value (but still well-defined).

tianchen92 replied:
Hmm，here is not "AppendEmpty" API, just move null_bit_map pointer to avoid set child arrays null when the top-level struct slot is null. I renamed to "Forward" here.

pitrou commented on cpp/src/arrow/array/builder_base.h:
```diff
@@ -98,6 +98,9 @@ class ARROW_EXPORT ArrayBuilder {
   virtual Status AppendNull() = 0;
   virtual Status AppendNulls(int64_t length) = 0;
 
+  virtual Status AppendEmpty() { return AppendNull(); }
```
Well, we'd like AppendEmpty to precisely append a (arbitrary but defined) non-null value, so this is not correct.

tianchen92 replied:
done

emkornfield commented:
@pitrou does this seem ok now?

pitrou commented on cpp/src/arrow/pretty_print.cc:
```diff
@@ -339,6 +339,17 @@ class ArrayPrinter : public PrettyPrinter {
+      // because StructBuilder#AppendNull dosen't set child null bitmap.
+      auto field_null_buf = array.field(i)->null_bitmap();
+      if (field_null_buf != nullptr) {
+        uint8_t* field_null_data = field_null_buf->mutable_data();
```
Why exactly are you mutating a null bitmap in a pretty printer?

tianchen92 replied:
Thanks for review, I have some questions, correct me if I am wrong:
1 now seems AppendNull/AppendNulls set null bitmap and append empty value using data_builder_, if AppendEmpty/AppendEmpties also do this, what is the difference between this two API?
2 From http://arrow.apache.org/docs/format/Columnar.html#struct-layout that there is no need for the child arrays to be null when the top-level struct slot is null, so I thought only 'empty' value should be appended (without setting null bitmap)。So in PrettyPrint, we need to change child null bitmap to ensure the correct output.

pitrou replied:
The difference should be this:
* `AppendNull` appends a null entry in the struct
* `AppendEmptyValue` appends a non-null entry in the struct

pitrou replied:
> So in PrettyPrint, we need to change child null bitmap to ensure the correct output

We certainly must not *change* the null bitmap. PrettyPrint just shows the child arrays as they are, it does not need to be changed here (such arrays can already be produced otherwise).

tianchen92 replied:
> The difference should be this:
> 
> * `AppendNull` appends a null entry in the struct
> * `AppendEmptyValue` appends a non-null entry in the struct

Thanks for your clarification, agree with you and I have updated this PR, please take a look, thanks!

tianchen92 replied:
> > So in PrettyPrint, we need to change child null bitmap to ensure the correct output
> 
> We certainly must not _change_ the null bitmap. PrettyPrint just shows the child arrays as they are, it does not need to be changed here (such arrays can already be produced otherwise).

Ok, reverted.

pitrou commented on cpp/src/arrow/array/builder_nested.h:
```diff
@@ -109,6 +109,25 @@ class BaseListBuilder : public ArrayBuilder {
   }
 
+  Status AppendEmpty() final {
+    null_bitmap_builder_.Forward(1);
```
Shouldn't you `Reserve(1)`?

tianchen92 replied:
Done

pitrou commented on cpp/src/arrow/array/builder_nested.h:
```diff
@@ -109,6 +109,25 @@ class BaseListBuilder : public ArrayBuilder {
+  }
+
+  Status AppendEmpties(int64_t length) final {
+    ARROW_RETURN_NOT_OK(CheckNextOffset());
```
Shouldn't you `Reserve(length)`?

tianchen92 replied:
Done

pitrou commented on cpp/src/arrow/buffer_builder.h:
```diff
@@ -292,6 +292,11 @@ class TypedBufferBuilder<bool> {
     return Status::OK();
   }
 
+  void Forward(int64_t num_elements) {
```
This will leave some entries undefined unless they're written manually elsewhere, which I don't think happens here.
I don't think this method is needed, instead you should just `Append(true)` or `Append(length, true)`, IMHO.

tianchen92 replied:
See comments above~

tianchen92 replied:
This method was removed.

pitrou commented:
As for naming, according to various dictionaries, "empty" as a noun (and "empties") is rather informal.
It seems `AppendEmptyValue` and `AppendEmptyValues` may be more appropriate (and more explicit), but I'm not a native English speaker.

tianchen92 commented:
> As for naming, according to various dictionaries, "empty" as a noun (and "empties") is rather informal.
> It seems `AppendEmptyValue` and `AppendEmptyValues` may be more appropriate (and more explicit), but I'm not a native English speaker.

Fine, fixed.

tianchen92 commented:
Hi, @emkornfield , please also take a look at this patch when you have time.
This patch add AppendEmptyValue/AppendEmptyValues used in StructBuilder#AppendNull/AppendNulls, which would set child null bitmap slot to true rather than false( http://arrow.apache.org/docs/format/Columnar.html#struct-layout).
And now this test ArrowReaderWriter#SingleColumnNullableStruct failed, could you please help fix this since you're export in parquet :) Thanks very much!  error message as below:

>>Failed
Unequal at absolute position 2
@@ -1, +1 @@
-{Backward: 0}
+{Backward: 10}
Expected:
  -- is_valid:
      [
      false,
      true
    ]
  -- child 0 type: int64
    [
      0,
      10
    ]
Actual:
  -- is_valid:
      [
      false,
      true
    ]
  -- child 0 type: int64
    [
      null,
      0
    ]
[  FAILED  ] ArrowReadWrite.SingleColumnNullableStruct (7 ms)

pitrou commented on cpp/src/arrow/array/builder_primitive.h:
```diff
@@ -43,6 +43,14 @@ class ARROW_EXPORT NullBuilder : public ArrayBuilder {
 
+  Status AppendEmptyValues(int64_t length) final {
+    if (length < 0) return Status::Invalid("length must be positive");
+    length_ += length;
```
For the null type, `AppendEmptyValues` should simply be the same as `AppendNulls` (the only possible value is null).

pitrou commented on cpp/src/arrow/json/parser_test.cc:
```diff
@@ -188,7 +189,7 @@ TEST(BlockParserWithSchema, Nested) {
                      {"[\"thing\", null, \"\xe5\xbf\x8d\", null]",
                       R"([["1", "2", "3"], ["2"], [], null])",
-                      R"([{"ps":null}, null, {"ps":"78"}, {"ps":"90"}])"});
+                      R"([{"ps":null}, {}, {"ps":"78"}, {"ps":"90"}])"});
```
Hmm... @bkietz Can you validate this change is ok? It seems it only affects internal implementation details.

bkietz replied:
This seems fine

pitrou reviewed (commented):
Thanks for doing this. It looks a lot better now! Here are some more comments.

pitrou commented:
@tianchen92 The Parquet failure is probably a bug in the nested Parquet reader. I've pushed changes to your branch, be careful to pick them up.

emkornfield commented:
@tianchen92 sorry I missed the tag.  Is it possible the failure could be related to https://issues.apache.org/jira/browse/ARROW-9603?filter=-2

tianchen92 commented:
@pitrou Thanks for your patch and careful review, I updated PR according to the comments above.
@emkornfield doesn't matter :), pitrou helped disabled the failed test and now seems unblocked.

emkornfield commented on cpp/src/arrow/array/builder_base.h:
```diff
@@ -98,6 +98,9 @@ class ARROW_EXPORT ArrayBuilder {
   virtual Status AppendNull() = 0;
   virtual Status AppendNulls(int64_t length) = 0;
 
+  virtual Status AppendEmptyValue() = 0;
```
these could really use some documentation it isn't entirely clear to me there purpose, and why they are separate from append null and what functionality ends up changing by call this in structbuilder.AppendNull

tianchen92 replied:
I add documentation for this API. The change meets the [spec](http://arrow.apache.org/docs/format/Columnar.html#struct-layout)(There is no need for the child arrays to be null when the top-level struct slot is null) .
The difference between append empty and append null value is different.

emkornfield replied:
Yes, the change meets the spec but potentially breaks parquet writing for use-cases going through th builder per https://issues.apache.org/jira/browse/ARROW-9603?filter=-2 (which relies on the behavior of forcing nulls).  I would prefer not to change this behavior before fixing the referenced bug.

tianchen92 replied:
I am ok, let's hold this on.

emkornfield replied:
the fix for arrow 9603 has been merged, could we try reverting change for parquet and seeing what passes/fails.

nealrichardson commented:
@tianchen92 do you want to pick this up again? the bug that @emkornfield mentioned has been fixed.

tianchen92 commented:
@emkornfield @nealrichardson @pitrou Sorry for the delay due to National Day Holiday, I reverted parquet related changes and rebased this PR. Seems tests all passed (one unrelated failing). Please help take a look again, thanks!

emkornfield commented:
@pitrou any objections to merging now (I would expect this might conflict with Decimal256 PR).

pitrou commented:
Working on this.

pitrou reviewed (approved):
Added some fixes and tests. +1.
